e2e/nomostest/registryproviders/artifact_registry.go

```diff
@@ -0,0 +1,179 @@
+	project string
+	// location to store the image
+	location string
+	// repositoryName in which to store the image
+	repositoryName string
+	// repositorySuffix is a suffix appended after the repository name. This enables
+	// separating different artifact types within the repository
+	// (e.g. helm chart vs basic oci image). This is not strictly necessary but
+	// provides a guardrail to help prevent gotchas when mixing usage of helm-sync
+	// and oci-sync.
+	repositorySuffix string
+	// shell used for invoking CLI tools
+	shell *testshell.TestShell
+}
+
+// Type returns the provider type.
+func (a *ArtifactRegistryProvider) Type() string {
+	return e2e.ArtifactRegistry
+}
+
+// Teardown preforms teardown
+// Does nothing currently. We may want to have this delete the repository to
+// minimize side effects.
+func (a *ArtifactRegistryProvider) Teardown() error {
+	return nil
+}
+
+// registryHost returns the domain of the artifact registry
+func (a *ArtifactRegistryProvider) registryHost() string {
+	return fmt.Sprintf("%s-docker.pkg.dev", a.location)
+}
+
+// repositoryAddress returns the domain and path to the chart repository
+func (a *ArtifactRegistryProvider) repositoryAddress() string {
+	return fmt.Sprintf("%s/%s/%s/%s", a.registryHost(), a.project, a.repositoryName, a.repositorySuffix)
+}
+
+// createRepository uses gcloud to create the repository, if it doesn't exist.
+func (a *ArtifactRegistryProvider) createRepository() error {
+	out, err := a.shell.ExecWithDebug("gcloud", "artifacts", "repositories",
+		"describe", a.repositoryName,
+		"--location", a.location,
+		"--project", a.project)
+	if err != nil {
+		if !strings.Contains(string(out), "NOT_FOUND") {
+			return fmt.Errorf("failed to describe image repository: %w", err)
+		}
+		// repository does not exist, continue with creation
+	} else {
+		// repository already exists, skip creation
+		return nil
+	}
+
+	_, err = a.shell.ExecWithDebug("gcloud", "artifacts", "repositories",
+		"create", a.repositoryName,
+		"--repository-format", "docker",
+		"--location", a.location,
+		"--project", a.project)
+	if err != nil {
+		return fmt.Errorf("failed to create image repository: %w", err)
+	}
+	return nil
+}
+
+// deleteImage the package from the remote registry, including all versions and tags.
+func (a *ArtifactRegistryProvider) deleteImage(name, digest string) error {
+	imageURL := fmt.Sprintf("%s/%s@%s", a.repositoryAddress(), name, digest)
+	if _, err := a.shell.ExecWithDebug("gcloud", "artifacts", "docker", "images", "delete", imageURL, "--delete-tags", "--project", a.project); err != nil {
+		return fmt.Errorf("deleting image from registry: %w", err)
+	}
+	return nil
+}
+
+// ArtifactRegistryOCIProvider provides methods for interacting with the test registry-server
+// using the oci-sync interface.
+type ArtifactRegistryOCIProvider struct {
+	ArtifactRegistryProvider
+}
+
+// Setup performs setup
+func (a *ArtifactRegistryOCIProvider) Setup() error {
```

Comment:
The Helm provider sets up the authentication and create the repository.
The OCI provider also needs to set up the authentication.

Reply:
Done